File: core/src/main/java/dev/vibris/core/request/RequestRegistry.java
package dev.vibris.core.request;

import java.time.Clock;
import java.time.Duration;
import java.time.Instant;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;

public final class RequestRegistry<T> {
    private final int liveCapacity;
    private final int terminalCapacity;
    private final Duration terminalTtl;
    private final Clock clock;
    private final Map<String, Entry<T>> live = new LinkedHashMap<>();
    private final Map<String, Entry<T>> terminal = new LinkedHashMap<>();

    public RequestRegistry(int liveCapacity, int terminalCapacity, Duration terminalTtl, Clock clock) {
        if (liveCapacity < 1 || terminalCapacity < 1) {
            throw new IllegalArgumentException("request registry capacities must be positive");
        }
        if (terminalTtl.isNegative() || terminalTtl.isZero()) {
            throw new IllegalArgumentException("terminal request TTL must be positive");
        }
        this.liveCapacity = liveCapacity;
        this.terminalCapacity = terminalCapacity;
        this.terminalTtl = terminalTtl;
        this.clock = Objects.requireNonNull(clock, "clock");
    }

    public synchronized AcceptResult<T> accept(String requestId) {
        return accept(requestId, "");
    }

    public synchronized AcceptResult<T> accept(String requestId, String ownerId) {
        requireId(requestId);
        Objects.requireNonNull(ownerId, "ownerId");
        purgeExpired();
        Entry<T> current = live.get(requestId);
        if (current != null) {
            if (!current.ownerId.equals(ownerId)) return new AcceptResult<>(AcceptKind.OWNER_MISMATCH, null);
            return new AcceptResult<>(AcceptKind.CURRENT, current.snapshot());
        }
        Entry<T> completed = terminal.get(requestId);
        if (completed != null) {
            if (!completed.ownerId.equals(ownerId)) return new AcceptResult<>(AcceptKind.OWNER_MISMATCH, null);
            return new AcceptResult<>(AcceptKind.CACHED_FINAL, completed.snapshot());
        }
        if (live.size() == liveCapacity) {
            return new AcceptResult<>(AcceptKind.FULL, null);
        }
        Entry<T> accepted = new Entry<>(ownerId, RequestState.ACCEPTED, null, null);
        live.put(requestId, accepted);
        return new AcceptResult<>(AcceptKind.NEW, accepted.snapshot());
    }

    public synchronized void markRunning(String requestId) {
        Entry<T> entry = requireLive(requestId);
        if (entry.state != RequestState.ACCEPTED) {
            throw new IllegalStateException("request is not accepted: " + requestId);
        }
        entry.state = RequestState.RUNNING;
    }

    public synchronized void finish(String requestId, RequestState state, T result) {
        if (!Objects.requireNonNull(state, "state").terminal()) {
            throw new IllegalArgumentException("final request state must be terminal");
        }
        Entry<T> entry = requireLive(requestId);
        live.remove(requestId);
        entry.state = state;
        entry.result = Objects.requireNonNull(result, "result");
        entry.completedAt = clock.instant();
        purgeExpired();
        while (terminal.size() >= terminalCapacity) {
            terminal.remove(terminal.keySet().iterator().next());
        }
        terminal.put(requestId, entry);
    }

    public synchronized Optional<Snapshot<T>> resume(String requestId) {
        return resume(requestId, "");
    }

    public synchronized Optional<Snapshot<T>> resume(String requestId, String ownerId) {
        requireId(requestId);
        Objects.requireNonNull(ownerId, "ownerId");
        purgeExpired();
        Entry<T> entry = live.get(requestId);
        if (entry == null) entry = terminal.get(requestId);
        if (entry != null && !entry.ownerId.equals(ownerId)) return Optional.empty();
        return entry == null ? Optional.empty() : Optional.of(entry.snapshot());
    }

    public synchronized int liveSize() {
        return live.size();
    }

    public synchronized int terminalSize() {
        purgeExpired();
        return terminal.size();
    }

    public synchronized int size() {
        purgeExpired();
        return live.size() + terminal.size();
    }

    private Entry<T> requireLive(String requestId) {
        requireId(requestId);
        Entry<T> entry = live.get(requestId);
        if (entry == null) throw new IllegalStateException("unknown live request: " + requestId);
        return entry;
    }

    private void purgeExpired() {
        Instant cutoff = clock.instant().minus(terminalTtl);
        terminal.entrySet().removeIf(entry -> !entry.getValue().completedAt.isAfter(cutoff));
    }

    private static void requireId(String requestId) {
        if (requestId == null || requestId.isBlank()) {
            throw new IllegalArgumentException("request ID must not be blank");
        }
    }

    public enum AcceptKind {
        NEW,
        CURRENT,
        CACHED_FINAL,
        OWNER_MISMATCH,
        FULL
    }

    public record AcceptResult<T>(AcceptKind kind, Snapshot<T> snapshot) {
    }

    public record Snapshot<T>(RequestState state, T result) {
    }

    private static final class Entry<T> {
        private final String ownerId;
        private RequestState state;
        private T result;
        private Instant completedAt;

        private Entry(String ownerId, RequestState state, T result, Instant completedAt) {
            this.ownerId = ownerId;
            this.state = state;
            this.result = result;
            this.completedAt = completedAt;
        }

        private Snapshot<T> snapshot() {
            return new Snapshot<>(state, result);
        }
    }
}

File: core/src/main/kotlin/dev/vibris/core/request/RequestRegistry.kt
package dev.vibris.core.request

import java.time.Clock
import java.time.Duration
import java.time.Instant
import java.util.LinkedHashMap
import java.util.Optional

class RequestRegistry<T : Any>(
    private val liveCapacity: Int,
    private val terminalCapacity: Int,
    private val terminalTtl: Duration,
    private val clock: Clock,
) {
    private val live = LinkedHashMap<String, Entry<T>>()
    private val terminal = LinkedHashMap<String, Entry<T>>()

    init {
        require(liveCapacity >= 1 && terminalCapacity >= 1) {
            "request registry capacities must be positive"
        }
        require(!terminalTtl.isNegative && !terminalTtl.isZero) {
            "terminal request TTL must be positive"
        }
    }

    @Synchronized
    fun accept(requestId: String?): AcceptResult<T> = accept(requestId, "")

    @Synchronized
    fun accept(requestId: String?, ownerId: String?): AcceptResult<T> {
        val id = requireId(requestId)
        val owner = ownerId ?: throw NullPointerException("ownerId")
        purgeExpired()
        val current = live[id]
        if (current != null) {
            if (current.ownerId != owner) {
                return AcceptResult(AcceptKind.OWNER_MISMATCH, null)
            }
            return AcceptResult(AcceptKind.CURRENT, current.snapshot())
        }
        val completed = terminal[id]
        if (completed != null) {
            if (completed.ownerId != owner) {
                return AcceptResult(AcceptKind.OWNER_MISMATCH, null)
            }
            return AcceptResult(AcceptKind.CACHED_FINAL, completed.snapshot())
        }
        if (live.size == liveCapacity) {
            return AcceptResult(AcceptKind.FULL, null)
        }
        val accepted = Entry<T>(owner, RequestState.ACCEPTED)
        live[id] = accepted
        return AcceptResult(AcceptKind.NEW, accepted.snapshot())
    }

    @Synchronized
    fun markRunning(requestId: String?) {
        val entry = requireLive(requestId)
        check(entry.state == RequestState.ACCEPTED) {
            "request is not accepted: $requestId"
        }
        entry.state = RequestState.RUNNING
    }

    @Synchronized
    fun finish(requestId: String?, state: RequestState, result: T) {
        require(state.terminal()) { "final request state must be terminal" }
        val id = requireId(requestId)
        val entry = requireLive(id)
        live.remove(id)
        entry.state = state
        entry.result = result
        entry.completedAt = clock.instant()
        purgeExpired()
        while (terminal.size >= terminalCapacity) {
            terminal.remove(terminal.keys.iterator().next())
        }
        terminal[id] = entry
    }

    @Synchronized
    fun resume(requestId: String?): Optional<Snapshot<T>> = resume(requestId, "")

    @Synchronized
    fun resume(requestId: String?, ownerId: String?): Optional<Snapshot<T>> {
        val id = requireId(requestId)
        val owner = ownerId ?: throw NullPointerException("ownerId")
        purgeExpired()
        val entry = live[id] ?: terminal[id]
        if (entry != null && entry.ownerId != owner) {
            return Optional.empty()
        }
        return if (entry == null) Optional.empty() else Optional.of(entry.snapshot())
    }

    @Synchronized
    fun liveSize(): Int = live.size

    @Synchronized
    fun terminalSize(): Int {
        purgeExpired()
        return terminal.size
    }

    @Synchronized
    fun size(): Int {
        purgeExpired()
        return live.size + terminal.size
    }

    private fun requireLive(requestId: String?): Entry<T> {
        val id = requireId(requestId)
        return live[id] ?: throw IllegalStateException("unknown live request: $id")
    }

    private fun purgeExpired() {
        val cutoff = clock.instant().minus(terminalTtl)
        terminal.entries.removeIf { !it.value.completedAt!!.isAfter(cutoff) }
    }

    private fun requireId(requestId: String?): String {
        require(!requestId.isNullOrBlank()) { "request ID must not be blank" }
        return requestId
    }

    enum class AcceptKind {
        NEW,
        CURRENT,
        CACHED_FINAL,
        OWNER_MISMATCH,
        FULL,
    }

    @JvmRecord
    data class AcceptResult<T>(val kind: AcceptKind, val snapshot: Snapshot<T>?)

    @JvmRecord
    data class Snapshot<T>(val state: RequestState, val result: T?)

    private class Entry<T>(
        val ownerId: String,
        var state: RequestState,
        var result: T? = null,
        var completedAt: Instant? = null,
    ) {
        fun snapshot(): Snapshot<T> = Snapshot(state, result)
    }
}